### PR TITLE
Use optional chunksizes argument in register_restart_field calls

### PR DESCRIPTION
**Pull request by DusanJovic-NOAA**
**Description**

This PR updates the register_restart_field calls to pass optional argument 'chunksizes'. Chunk sizes are equal to the corresponding dimension size. This change will speed up reading the restart in RRFS DA.

Fixes https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/issues/217

**How Has This Been Tested?**

UFS weather model regression testing on Hera.

**Checklist:**

Please check all whether they apply or not
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by bensonr — COMMENTED** [review]
There are two potential issues here.   First, the chunksize specifications seem to only allow one to use the full domain for the file.  This may not be ideal for parallel reads done via sections.  Has this also been tested with an io_domain not (1,1)?

Second, by hardcoding in a default netcdf4 format for the restarts, this file will require special merging as science updates and other bugfixes are released via the main branch.

**Comment by DusanJovic-NOAA** [comment]
The chunk sizes will be equal to each dimension sizes, not the full domain. I will run the tests with an io_domain not equal to (1,1) to check if it works, I do not think we have such test in ufs-weather-model regression test.

**Comment by DusanJovic-NOAA** [comment]
I ran 'control' and 'control_restart' tests with io_layout(1,1) and io_layout(3,2). Here are netcdf dumps showing chunksizes.

Restart file chunksizes with io_layout(1,1):
```
$ ncdump -h -s /scratch1/NCEPDEV/stmp2/Dusan.Jovic/FV3_RT/rt_80826/control/RESTART/fv_core.res.tile1.nc
netcdf fv_core.res.tile1 {
dimensions:
        xaxis_1 = 96 ;
        xaxis_2 = 97 ;
        yaxis_1 = 97 ;
        yaxis_2 = 96 ;
        zaxis_1 = 127 ;
        Time = UNLIMITED ; // (1 currently)
variables:
        ...
        float u(Time, zaxis_1, yaxis_1, xaxis_1) ;
                u:checksum = "   1F0F0C040420E" ;
                u:_Storage = "chunked" ;
                u:_ChunkSizes = 1, 127, 97, 96 ;
                u:_Endianness = "little" ;
        float v(Time, zaxis_1, yaxis_2, xaxis_2) ;
                v:checksum = "    C8BECD700566" ;
                v:_Storage = "chunked" ;
                v:_ChunkSizes = 1, 127, 96, 97 ;
                v:_Endianness = "little" ;
        float W(Time, zaxis_1, yaxis_2, xaxis_1) ;
                W:checksum = "FFFB5271660F5152" ;
                W:_Storage = "chunked" ;
                W:_ChunkSizes = 1, 127, 96, 96 ;
                W:_Endianness = "little" ;
        ...
}
```

First restart file (out of 6) chunksizes with io_layout(3,2):
```
$ ncdump -h -s /scratch1/NCEPDEV/stmp2/Dusan.Jovic/FV3_RT/rt_144919/control/RESTART/fv_core.res.tile1.nc.0000
netcdf fv_core.res.tile1.nc {
dimensions:
        xaxis_1 = 32 ;
        xaxis_2 = 33 ;
        yaxis_1 = 49 ;
        yaxis_2 = 48 ;
        zaxis_1 = 127 ;
        Time = UNLIMITED ; // (1 currently)
variables:
        ...
        float u(Time, zaxis_1, yaxis_1, xaxis_1) ;
                u:checksum = "   1F0F0C040420E" ;
                u:_Storage = "chunked" ;
                u:_ChunkSizes = 1, 127, 49, 32 ;
                u:_Endianness = "little" ;
        float v(Time, zaxis_1, yaxis_2, xaxis_2) ;
                v:checksum = "    C8BECD700566" ;
                v:_Storage = "chunked" ;
                v:_ChunkSizes = 1, 127, 48, 33 ;
                v:_Endianness = "little" ;
        float W(Time, zaxis_1, yaxis_2, xaxis_1) ;
                W:checksum = "FFFB5271660F5152" ;
                W:_Storage = "chunked" ;
                W:_ChunkSizes = 1, 127, 48, 32 ;
                W:_Endianness = "little" ;
        ...
}
```

**Comment by jswhit** [comment]
It would also be nice to specify deflate_level and shuffle.  Default values for these parameters were settable in the old mpp_io namelist (and the GFS uses them to get compressed restart files).  This doesn't seem possible with fms2_io.

**Comment by DusanJovic-NOAA** [comment]
Where should the value for deflate_level and shuffle be specified? In fms2_io_nml, or some dycore namelist?

**Comment by DusanJovic-NOAA** [comment]
I see that register_restart_field function only has deflate_level as an optional argument, not shuffle.

File: tools/fv_io.F90
Comment (uramirez8707) [inline tools/fv_io.F90:835]: Instead of hard codding `netcdf4` here, you can set this namelist
https://github.com/NOAA-GFDL/FMS/blob/d8208e61be2dff46cc387cbace0f1f7c2bd950fc/fms2_io/fms2_io.F90#L382-L384

**Comment by uramirez8707** [comment]
> I see that register_restart_field function only has deflate_level as an optional argument, not shuffle.

That is correct, please open an issue in FMS so that argument can be added.

File: tools/fv_io.F90
Comment (DusanJovic-NOAA) [reply tools/fv_io.F90:835]: What will happen if user forgets to set that namelist parameter, for whatever reason, will FMS just ignore chunksizes argument or report an error and terminate the program?

Comment (uramirez8707) [reply tools/fv_io.F90:835]: It will output a warning and ignore them.
https://github.com/NOAA-GFDL/FMS/blob/main/fms2_io/netcdf_io.F90#L951

Comment (DusanJovic-NOAA) [reply tools/fv_io.F90:835]: In netcdf_io.F90, here:

https://github.com/NOAA-GFDL/FMS/blob/073c6003b90099c3b02084b087601ff3109e9f83/fms2_io/netcdf_io.F90#L621

`fileobj%is_netcdf4` is set to `.true.` only based on `nc_format` function argument, not based on namelist parameter ater line [629](https://github.com/NOAA-GFDL/FMS/blob/073c6003b90099c3b02084b087601ff3109e9f83/fms2_io/netcdf_io.F90#L629).

**Comment by jswhit** [comment]
> Where should the value for deflate_level and shuffle be specified? In fms2_io_nml, or some dycore namelist?

I think it would make the most sense to be able to set the default values for deflate_level and shuffle in fms2_io_nml, just like they were in mpp_io_nml.

File: tools/fv_io.F90
Comment (uramirez8707) [reply tools/fv_io.F90:835]: ahh this is a bug, thanks for letting us know. We will get this fixed!

**Comment by uramirez8707** [comment]
> I think it would make the most sense to be able to set the default values for deflate_level and shuffle in fms2_io_nml, just like they were in mpp_io_nml.

@DusanJovic-NOAA @jswhit The `deflate_level` and `shuffle` will be added to fms2_io_nml instead of being arguments to `register_restart_variables`. I think this will make it easier for you to use.

**Comment by jkbk2004** [comment]
Downstream [ufs wm pr 1467](https://github.com/ufs-community/ufs-weather-model/pull/1467) was fully tested ok thru [ufs wm pr 1497](https://github.com/ufs-community/ufs-weather-model/pull/1497). @bensonr @uramirez8707 @laurenchilutti Please, go ahead to merge this pr.

**Comment by jkbk2004** [comment]
@laurenchilutti @bensonr @uramirez8707 Just to remind we are waiting for this pr to be merged.

**Comment by bensonr** [comment]
There is a bug in FMS preventing the original comments in #223 from being addressed, namely hardcoded values of netcdf4 for the file types.  The next release of FMS should address this bug.  

GFDL would prefer not to accept and merge this as is, unless there is an immediate PR removing the hardcoded values for when FMS is ready.

**Comment by junwang-noaa** [comment]
@laurenchilutti @bensonr I saw the FMS PR #1075 was committed, so now we can specify deflate_level and shuffle in fms2_io_nml. May I ask if you have a timeline on when the new FMS release with this capability will be available? We'd like to use this feature in UFS applications. Thank you!

**Comment by laurenchilutti** [comment]
@junwang-noaa FMS 2023.01 (our next release which will have this update) is scheduled to be released approximately Feb 17th.  To give some insight on our release process: FMS uses alpha and beta tags during pre-release testing.  This is an approximate timeline of the pre-release tags and the final release:
- We (FMS developers) have already tested 2023.01-alpha1
- FMS was tagged with [2023.01-alpha2](https://github.com/NOAA-GFDL/FMS/releases/tag/2023.01-alpha2) today. We will test this over the next week
-  2023.01-alpha3 will come approx. Jan 27 with one week of testing to follow.
-  2023.01-beta1 will come approx. Feb 3.  We will then test a beta tag for about 2 weeks -- this testing is end to end testing and more thorough.
-   Final Release approx. Feb 17

You are welcome to begin testing with the alpha2 tag immediately if you would like to verify the updates are as you expect.

**Comment by junwang-noaa** [comment]
Thanks, Lauren. We can test the 2023.01-alpha2 to see if it works for us.

**Comment by DusanJovic-NOAA** [comment]
> Thanks, Lauren. We can test the 2023.01-alpha2 to see if it works for us.

@junwang-noaa I compiled 2023.01-alpha2 on Hera successfully and ran one regression test using this branch, it worked.

**Comment by hu5970** [comment]
Can we have an option to set:
All the fields have chunksizes with horizontal dimensions (nx,ny,1)?

This will be  good for parallel_netcdf options for many following up application, especially for huge domain case like NA 3km grid.

**Review by uramirez8707 — APPROVED** [review]

**Review by bensonr — APPROVED** [review]

**Comment by jkbk2004** [comment]
@DusanJovic-NOAA can we confirm if additional update is needed regarding deflate_level/shuffle parameters in fms2_io_nml ? what about hard coding issue of netcdf4 in open_file?

**Comment by DusanJovic-NOAA** [comment]
> @DusanJovic-NOAA can we confirm if additional update is needed regarding deflate_level/shuffle parameters in fms2_io_nml ? what about hard coding issue of netcdf4 in open_file?

I removed hard-coded netcdf4 option in https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/223/commits/28417df018724d7b4623c82615058b8589fca069

**Comment by zach1221** [comment]
@laurenchilutti testing is complete on UFS-WM PR #1467. Can you please review and merge this PR?

**Pull request merged**
